### DISCUSSION TITLE
Autopause++ (process hibernation or container management)

### DISCUSSION THREAD

**Discussion by Victoare** [itzg/docker-minecraft-server#2366] [General]
Hi all.

I have a small home server with a couple minecraft worlds playing with my family. Most of the time they are just sitting there auto-paused, consuming 2G memory each. Wondering if it is technically possible to hibernate the server process as such windows hibernate its whole system, writing out the memory content to a file and read back when the server is waking up.
I know (if it would be possible) there would be a much greater startup delay, but would be more resource friendly (at least in my case).

Another way would be to automatically shut down / restart the container.
In my case I use mc_router as well and maybe that or a similar container could monitor each instance (and client connections) and start/stop the containers accordingly.

**Comment by itzg** [comment]
I think I was reading docker has or it will have a feature to suspend a container with its memory state to disk. It would be nice to have mc-router invoke an API on docker to manage that.

**Comment by Victoare** [comment]
Found this docker command: https://docs.docker.com/reference/cli/docker/checkpoint which will save the memory to disk and load back from there on startup. Although it is an experimental feature for now, but maybe worth a shot.

**Comment by vaaski** [comment]
Would it be possible to shut down the server completely and restart it when someone tries to connect? Perhaps showing them a message that the server is currently starting? Maybe even a server MOTD that the server is currently hibernating?

**Reply by itzg** [reply]
Please take a look at one of these solutions https://docker-minecraft-server.readthedocs.io/en/latest/misc/examples/#lazymc-put-your-minecraft-server-to-rest-when-idle

**Reply by Victoare** [reply]
Ohh, nice! Thanks!
A little bit more convoluted that I imagined, but this seems to be doing what I thought at the first place.

Edit: Maybe a bit slower to actually shut down and restart the server (as this lazymc does) than hibernate the container with checkpoint, but maybe this is a more foolproof solution (who knows what would the mc server do if it sees a large time-jump after getting back from hibernation)

**Reply by vaaski** [reply]
That's exactly what I've been looking for and it works great, thanks!

Off-topic: I have two separate minecraft servers at the moment, but my machine can't handle both simultaneously. Is there an intended way to ensure that server B can't be started while server A is already running and active?

**Reply by itzg** [reply]
> Off-topic: I have two separate minecraft servers at the moment, but my machine can't handle both simultaneously. Is there an intended way to ensure that server B can't be started while server A is already running and active?

Technically using Kubernetes, such as one of the lightweight versions, could solve this since it'll only schedule pods that fit in the given cpu/memory resources of the node (machine).
